### PR TITLE
[New feature] Add flip-view and close buttons to pane-headers

### PR DESCRIPTION
**Pull request by petetnt**
This new feature improves the Splitscreen UX by adding buttons to panel headers that allow user to quickly flip the current view to another pane. It's not a complete solution, but another alternative to dragging the files over to another panel which can be a bit cumbersome, as evidenced in #11308, #11448, #11345 and others.

When Split View is enabled, arrow buttons are added to the panel that indicate flipping 

image

Clicking such button flips the view to the other panel, revealing the last used file in the pane it was flipped from

image

If there's no files open in the new other pane, the button obviously isn't there

image

This of course works in vertical split mode too:

image

The first commit doesn't include unit tests because Jasmine Spec Runner won't work for whatever reason: while I am looking at that issue, any feedback, improvements or questions are welcome.

~~edit: Build failed because the API rate limit exceeded~~

edit: 2
The panes now contain a `close buttons` too as similar to what @zaggino suggested.

image

Clicking the `x` closes the file, or prompts for confirmation if a file is dirty. If pane contains no files, the `close button` collapses the views by setting the layout scheme to `1x1`

### CODE REVIEW

**Comment by zaggino** [comment]
Love it, gave me an idea -> `X` mark in the corner to close split view quickly (one click instead of two), would solve my https://github.com/adobe/brackets/issues/11746 problem.

File: src/view/Pane.js [hunk 1]
```diff
@@ -159,17 +159,33 @@ define(function (require, exports, module) {
         EventDispatcher     = require("utils/EventDispatcher"),
         FileSystem          = require("filesystem/FileSystem"),
         InMemoryFile        = require("document/InMemoryFile"),
+        //WorkingSetView      = require("project/WorkingSetView"),
```
Comment (le717) [inline src/view/Pane.js]: Forget to remove a line?

File: src/styles/brackets.less [hunk 2]
```diff
@@ -486,11 +486,58 @@ a, img {
...
+            padding-right: 4px;
+            padding-bottom: 2px;
+            padding-left: 4px;
+            margin-left: 0;
+            margin-bottom: 0;
+            .sprite-icon(0, 0, 13px, 13px, "images/flip-view-icons.svg");
+            background-origin: content-box;
+            -webkit-transform: translateZ(0); // forces GPU mode for better filter rendering on retina
+            transform: translateZ(0); // future proofing
+            -webkit-filter: drop-shadow(0 1px 0 rgba(0,0,0,0.36));
```
Comment (le717) [inline src/styles/brackets.less]: `filter` has a standardized version. It'd probably be best to add that too like you did with `transform`.

Comment (petetnt) [inline src/styles/brackets.less]: Thanks @le717, fixed those :+1: 

RE: the filter declaration, there are some other components using the plain vendor-prefixed versions of ones that have standardized ones available, might be worth checking out.

**Comment by petetnt** [comment]
@zaggino yeah that is an great idea to me!

**Comment by petetnt** [comment]
Added an unit test

**Comment by petetnt** [comment]
The panes now contain a `close buttons` too as similar to what @zaggino suggested.

image

Clicking the `x` closes the file, or prompts for confirmation if a file is dirty. If pane contains no files, the `close button` collapses the views by setting the layout scheme to `1x1`

**Comment by zaggino** [comment]
few comments:
1) i like how the close highlights when hovered, can you make the same with the arrows? right now arrow icon doesn't highlight but some weird border around it highlights
2) the icons doesn't have to be there all the time, maybe then can appear only when I hover over the title area? (i mean the `.pane-header` element)
3) when i click close icon on the last file in the right panel, it doesn't close but i have to click again to close the empty panel?
4) if i have one file in the left panel and one file in the right panel, when i close the left file and then close the left pane, the right file disappears (it works fine closing the right panel)

**Comment by petetnt** [comment]
Thanks for the comments @zaggino. I'll look into those asap, esp. 4), not sure what's up with that :hammer: It seems that [`getCurrentlyViewedFile()`](https://github.com/petetnt/brackets/blob/petetnt/flip-view-buttons/src/view/Pane.js#L1190) returns the file in the active pane and just clicking on the `x` doesn't active the pane itself -> wrong file closes.

I was putting some behavior for 2) and 3) behind some prefs: `hover`, `always` and `off` for the panel buttons (default being `hover`) and something like `CLOSE_PANE_AFTER_LAST` for the panes:

**Comment by zaggino** [comment]
Ah right, I didn't read the code before testing this so I didn't knew about the various preferences. But show on `hover` seems to be reasonable default.

**Comment by petetnt** [comment]
@zaggino You didn't miss anything, the prefs haven't been implemented yet :)

**Comment by zaggino** [comment]
Ah right, keep the work coming then, it'd be great to see this in 1.6 :)

**Comment by petetnt** [comment]
Made additions and modifications according to @zaggino 's comments

- Unified hover styles for the flip view buttons
- Added preferences for showing the buttons and collapsing the pane after last file is closed
- Fixed bug where close button would close the wrong file if the panel was not active

File: src/view/Pane.js [hunk 3]
```diff
@@ -161,15 +161,44 @@ define(function (require, exports, module) {
...
+     * Internal pane id
+     * @const
+     * @private
+     */
+    var SECOND_PANE         = "second-pane";
+
+    // Define showPaneHeaderButtons, which controls when to show close and flip-view buttons
+    // on the header. Possible values "hover", "always" and "never"
+    PreferencesManager.definePreference("pane.showPaneHeaderButtons", "string", "hover", {
+        description: Strings.DESCRIPTION_SHOW_PANE_HEADER_BUTTONS
```
Comment (sprintr) [inline src/view/Pane.js]: You seems to have forgotten to add `"hover" "always"  "never"` to values array here so it can be shown in the hints.

Comment (petetnt) [inline src/view/Pane.js]: Thanks @sprintr , fixed that.

**Comment by abose** [comment]
@petetnt I can't see the close button when i tested this out. Do I need to do something more to enable close butoon?

**Comment by petetnt** [comment]
@abose good catch, the close button is there but you cannot see it when you are using a `light` theme such as Brackets default. Didn't test it out on a lighter theme after adding the close button, will fix asap

**Comment by petetnt** [comment]
@abose It's now fixed :hammer:

**Comment by abose** [comment]
works now :sunglasses: 
image
When a pane is empty, the flip button seems to be confused.

**Comment by petetnt** [comment]
@abose fixed that regression (nitpicky CSS inheriting :fire_engine:).

**Comment by abose** [comment]
Thanks for this feature @petetnt .
All unit/integration tests passing. LGTM.
When this is ok to be merged, Please remove the [review] tag from the title.

**Comment by petetnt** [comment]
Ready to be merged as far as I am concerned unless others have something else to add.

**Comment by abose** [comment]
Thanks @zaggino @sprintr @le717 for reviewing.
Merging.

**Comment by petetnt** [comment]
:fireworks: :tada: :+1:

**Comment by zaggino** [comment]
cool stuff, nice work @petetnt ;)

**Comment by arctwelve** [comment]
@petetnt Congrats on the new feature.

**Pull request merged**